[Messaging] Allow binary application properties

jsquire opened the pull request:
# Summary

The focus of these changes is to update AMQP conversion to remove the validation that disallows `byte[]` and `ArraySegment<byte>` values in application properties. These are confirmed to be encoded as [binary](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-binary) by `Microsoft.Azure.Amqp`, making them allowable values according to [section 3.2.5](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-application-properties) of the AMQP specification.

The Event Hubs documentation for `EventData` has been updated to reflect this change.  Also riding along is a non-trivial clean-up of the Event Hubs-specific AMQP message converter, which had a number of members that were replaced by the `Azure.Core.Amqp` converter implementation.

Due to a known bug in the Service Bus service, messages with `byte[]` application properties result in an error when being sent from the client.  Until this is confirmed fixed, the Service Bus documentation still reflects `byte[]` as an invalid type for the application properties.  While the client will no longer validate `byte[]`, any use will trigger a service error.  Regrettably, this does change the error behavior when this specific type is used, but this is both short-term (confirmed that the service bug will be fixed this semester) and it is enough of an edge case that I do not believe taking a performance hit on each serialization to maintain the current behavior is justified.  Likewise, I don't believe that this is impactful enough to take on the technical debt required to add a short-term flag to customize serialization behavior.

jsquire committed:
[Messaging] Allow binary application properties

The focus of these changes is to update AMQP conversion
to remove the validation that disallows `byte[]` and
`ArraySegment<byte>` values in application properties.
These are confirmed to be encoded as `binary` by
`Microsoft.Azure.Amqp`, making them allowable values
according to section 3.2.5 of the AMQP specification.

The Event Hubs documentation for `EventData` has been
updated to reflect this change.  Due to a known bug
in the Service Bus service, the service will reject
messages with `byte[]` application properties.  Until
this is confirmed fixed, the Service Bus documentation
still reflects `byte[]` as an invalid type for the
application properties.

christothes commented on sdk/eventhub/Azure.Messaging.EventHubs/src/EventData.cs:
```diff
@@ -243,6 +243,7 @@ public string CorrelationId
         ///     <item><description>Stream</description></item>
         ///     <item><description>Uri</description></item>
         ///     <item><description>TimeSpan</description></item>
+        ///     <item><description>byte[]</description></item>
```
Should we also add `ArraySegment<byte>`?

jsquire replied:
I left that off intentionally, because when you receive from the service, it gets transformed into `byte[]` in all cases because there is no way to differentiate between the types.   You can add an array segment when sending, but for all intents and purposes it becomes `byte[]`

jsquire committed:
Fixing EventData equality comparer used by tests to compare binary properties by sequence rather than simple equality.

jsquire commented:
/azp run net - eventhub - tests

jsquire committed:
Extending test cases for EventDataExtensionsTests
and adding docs for the service error related to
binary properties with suggested workaround.

m-redding reviewed (approved):
this looks good to me - the tests are very comprehensive 👍

christothes commented on sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Testing/EventDataExtensions.cs:
```diff
@@ -115,7 +115,53 @@ internal static class EventDataExtensions
+                // Properties can contain byte[] or ArraySegment<byte> values, which need to be compared
+                // as a sequence rather than by strict equality.
+
+                if (instance.Properties[key] is byte[] instanceByteArray)
+                {
+                    var otherByteArray = otherValue switch
+                    {
+                        byte[] byteArray => byteArray,
+                        ArraySegment<byte> arraySegment => arraySegment.ToArray(),
+                        _ => Array.Empty<byte>()
+                    };
+
+                    if (!Enumerable.SequenceEqual(instanceByteArray, otherByteArray))
+                    {
+                        return false;
+                    }
+                }
+                else if (instance.Properties[key] is ArraySegment<byte> instanceArraySegment)
+                {
+                    var otherArraySegment = otherValue switch
+                    {
+                        ArraySegment<byte> arraySegment => arraySegment,
+                        byte[] byteArray => new ArraySegment<byte>(byteArray),
+                        _ => new ArraySegment<byte>()
+                    };
+
+                    if (!Enumerable.SequenceEqual(instanceArraySegment, otherArraySegment))
+                    {
+                        return false;
+                    }
```
Could we simplify this and avoid possible `.ToArray` allocations by casting the values as `IList` or `ICollection` and comparing the sequences that way?

jsquire replied:
You know... I never realized that `Array<T>` implemented `IList<T>`.    Looks like we can do this.   Nice suggestion!

jsquire committed:
Simplify equality check
